### PR TITLE
GH-42168: [Python][Parquet] Pyarrow store decimal as integer

### PR DESCRIPTION
**Pull request by bkief**
This PR exposes the [store_decimal_as_integer](https://arrow.apache.org/docs/cpp/api/formats.html#_CPPv4N7parquet16WriterProperties7Builder31enable_store_decimal_as_integerEv) parquet writer property to pyarrow

### Rationale for this change

This will allow storing fixed-point decimals as integers in the parquet format and take advantage of more efficient storage codecs

### What changes are included in this PR?

Pyarrow parquet writer and related Cython 

### Are these changes tested?

Tests were included for the new parameter

### Are there any user-facing changes?

Docstrings were updated

* GitHub Issue: #42168

### CODE REVIEW

**Review by mapleFU — COMMENTED** [review]
General LGTM
BTW, You can follow some logic here: https://github.com/apache/arrow/pull/38360/files

**Comment by bkief** [comment]
@mapleFU or @AlenkaF - Can we merge this before the July 1  v17 cutoff?

File: python/pyarrow/_dataset_parquet.pyx
Comment (mapleFU) [inline python/pyarrow/_dataset_parquet.pyx:616]: why a `()` is added?

**Comment by mapleFU** [comment]
Sorry for the delay, I'll take a around and try that

File: python/pyarrow/_dataset_parquet.pyx
Comment (bkief) [reply python/pyarrow/_dataset_parquet.pyx:616]: I believe that was a linter/formatter recommendation, but its certainly not necessary

Comment (mapleFU) [reply python/pyarrow/_dataset_parquet.pyx:616]: The core and `test_basic.py` LGTM, but I'm not sure this part ( _dataset ) is tested, I'll check that tomorrow

File: python/pyarrow/_parquet.pyx
Comment (rok) [inline python/pyarrow/_parquet.pyx:1948]: Do we expect `store_decimal_as_integer` to ever not be a list of columns like `use_byte_stream_split` can be above? If not we don't need this check.

Comment (bkief) [reply python/pyarrow/_parquet.pyx:1948]: At present, I believe the C++ method does not accept columns.

Comment (rok) [reply python/pyarrow/_parquet.pyx:1948]: In that case do we need to check if it's a boolean or not?

Comment (rok) [reply python/pyarrow/_parquet.pyx:1948]: What I'm trying to say is we can probably remove the `if isinstance(store_decimal_as_integer, bool):` check if we don't need `store_decimal_as_integer` to ever be non-boolean.

**Review by mapleFU — APPROVED** [review]
General this LGTM as a C++ Parquet side menber, but I'm not familiar with Python, this requires help of other committers

File: python/pyarrow/_parquet.pyx
Comment (raulcd) [inline python/pyarrow/_parquet.pyx:1835]: Unnecessary empty line
```suggestion
```

File: python/pyarrow/_parquet.pyx
Comment (raulcd) [reply python/pyarrow/_parquet.pyx:1948]: If we expect this to be always a boolean, in my opinion, we probably should raise a `ValueError` if it is not a boolean, as we do with other options, example `column_encoding` :
```
        else:
            raise TypeError(
                "'column_encoding' should be a dictionary or a string")
```
instead of silently ignoring it.

**Review by raulcd — COMMENTED** [review]
in general looks good to me, some minor comments. 
@pitrou @jorisvandenbossche if you could take a quick look so we can add this to 17.0.0

File: python/pyarrow/_parquet.pyx
Comment (rok) [reply python/pyarrow/_parquet.pyx:1948]: As this is if `store_decimal_as_integer` is non-boolean it will get ignored. We could raise in case of other types to ensure hygiene?
Also isn't one of the two behaviors default? Why don't we just set in case we're using the non default one?
```
        if store_decimal_as_integer:
            props.enable_store_decimal_as_integer()
        else:
            props.disable_store_decimal_as_integer()
```

Comment (pitrou) [reply python/pyarrow/_parquet.pyx:1948]: @rok is right, ignoring arguments like this is very bad.

File: python/pyarrow/_dataset_parquet.pyx
Comment (bkief) [reply python/pyarrow/_dataset_parquet.pyx]: > why a `()` is added?

My latest commit removed this

File: python/pyarrow/_parquet.pyx
Comment (bkief) [reply python/pyarrow/_parquet.pyx:1835]: I believe this is resolved in the latest commit

File: python/pyarrow/_parquet.pyx
Comment (bkief) [reply python/pyarrow/_parquet.pyx:1948]: I added an `else` clause which raises a `TypeError`. 

I did not add a default props.*able_store_decimal_as_integer() as I thought it would be better to defer to the C++ writer defaults, which I believe is to disable.

File: python/pyarrow/_parquet.pyx
Comment (raulcd) [inline python/pyarrow/_parquet.pyx:1954]:
```suggestion
        raise TypeError("'store_decimal_as_integer' must be a boolean")
```

File: python/pyarrow/parquet/core.py
Comment (raulcd) [inline python/pyarrow/parquet/core.py:887]:
```suggestion
    When enabled, the writer will use the following physical types to store decimals:
```

**Review by raulcd — COMMENTED** [review]
A couple of minor grammar suggestions

File: python/pyarrow/_parquet.pyx
Comment (rok) [reply python/pyarrow/_parquet.pyx:1948]: >  I believe its better to be explicit rather than implicit here in the if/else with the property

That's a fair point - if c++ defaults change it would surprise python users.

File: python/pyarrow/parquet/core.py
Comment (bkief) [reply python/pyarrow/parquet/core.py:887]: I lazily copied and pasted from the Cpp docs. Shall we update that as well or is there a more suitable open PR for those changes?

**Review by rok — APPROVED** [review]
LGTM. Thanks for quick iterations @bkief !

**Review by wgtmac — APPROVED** [review]
+1

**Review by raulcd — COMMENTED** [review]
I have 0 experience on how we are testing parquet from Python. This PR seems to work and we can write and read to/from parquet when we set this new property.
The test does not validate that the stored value is indeed using int but I am not sure if this is already covered on the C++ tests.
I guess my only concern is that we are not really validating that the functionality is doing anything new, we only test we are able to set the flag and writing / reading.
That's why I can't give a +1 (will wait for others with more experience on this part of the code).

**Comment by mapleFU** [comment]
>  I guess my only concern is that we are not really validating that the functionality is doing anything new, we only test we are able to set the flag and writing / reading.

Maybe a way is check the parquet schema and verify the column schema is Integer ( https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ColumnSchema.html#pyarrow.parquet.ColumnSchema )

**Comment by raulcd** [comment]
> Maybe a way is check the parquet schema and verify the column schema is Integer ( https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ColumnSchema.html#pyarrow.parquet.ColumnSchema ).

That's a good idea, at least we can check that there is a difference on the schema between setting the flag to True or False.

**Comment by bkief** [comment]
> I have 0 experience on how we are testing parquet from Python. This PR seems to work and we can write and read to/from parquet when we set this new property. The test does not validate that the stored value is indeed using int but I am not sure if this is already covered on the C++ tests. I guess my only concern is that we are not really validating that the functionality is doing anything new, we only test we are able to set the flag and writing / reading. That's why I can't give a +1 (will wait for others with more experience on this part of the code).

I did consider directly inspecting the schema, and maybe this is better for directness, this but I found another way to test the same.  The [test with delta encoding](https://github.com/bkief/arrow/blob/pyarrow-store-decimal-as-integer/python/pyarrow/tests/parquet/test_basic.py#L379-L388) ensures the fields are stored as integers, if not the writer would throw an error as `DELTA_BINARY_PACKED` requires the columns to be of type `INT32` or `INT64`.

**Comment by raulcd** [comment]
> @raulcd I did consider directly inspecting the schema, and maybe this is better for directness, this but I found another way to test the same. The [test with delta encoding](https://github.com/bkief/arrow/blob/pyarrow-store-decimal-as-integer/python/pyarrow/tests/parquet/test_basic.py#L379-L388) ensures the fields are stored as integers, if not the writer would throw an error as `DELTA_BINARY_PACKED` requires the columns to be of type `INT32` or `INT64`.

Thanks! that sounds good to me, could you add a comment to the test to explain the above to make it more explicit?

**Comment by bkief** [comment]
I just went ahead and added the explicit test. It would be a good pattern to have for future tests

**Review by mapleFU — APPROVED** [review]

**Comment by mapleFU** [comment]
@raulcd should we better set this a blocker? I'll merge this if all CI pass and no negative comments in 6.30 night

**Comment by bkief** [comment]
> I just went ahead and added the explicit test. It would be a good pattern to have for future tests

Looks like there was some filesytem permission errors in the AppVeyor CI. I'll try a different pattern to see if that resolves it

**Comment by mapleFU** [comment]
CI failed is unrelated ( Maybe ORC related), merged

**Pull request merged**
